File: silicon/names.py
from __future__ import annotations

from typing import *
from silicon import ast
from silicon.diagnostics import *
from silicon.source import *
from dataclasses import dataclass, field

__all__ = ["resolve_names"]

BUILTIN_FUNCS = ("concat", "wire", "reg")


@dataclass
class Scope:
  parent: Optional[Scope]
  cannot_shadow: bool = False
  names: Dict[str, ast.AstNode] = field(default_factory=dict)

  # Resolve a `name` to an AST node in this scope or one of its parents.
  # Reports an error diagnostic for the given `loc` if the name is not found.
  def resolve(self, name: str, loc: Loc) -> ast.AstNode:
    scope: Optional[Scope] = self
    while scope:
      if node := scope.names.get(name):
        return node
      scope = scope.parent
    emit_error(loc, f"unknown name `{name}`")

  # Declare a name in this scope.
  def declare(self, name: str, node: ast.AstNode):
    if existing_node := self.names.get(name):
      if cannot_be_shadowed(existing_node) or self.cannot_shadow:
        emit_info(existing_node.loc,
                  f"previous definition of `{name}` was here")
        emit_error(node.loc, f"name `{name}` already defined")
    self.names[name] = node


# Returns true if the given node's name cannot be shadowed in a scope, and a
# "name already defined" error should be thrown.
def cannot_be_shadowed(node: ast.AstNode) -> bool:
  if isinstance(node, ast.ModItem):
    return True
  return False


# Resolve names in an AST.
def resolve_names(root: ast.Root):
  resolve_node(root, Scope(parent=None))

  # Ensure that no unresolved names remain in the AST.
  for child in root.walk():
    if isinstance(child,
                  ast.CallExpr) and child.name.spelling() in BUILTIN_FUNCS:
      continue
    for name, value in child.__dict__.items():
      if isinstance(value, ast.Binding) and value.target is None:
        emit_error(child.loc,
                   f"unresolved {name} in {child.__class__.__name__}")


# Resolve the names within the given node.
def resolve_node(node: ast.AstNode, scope: Scope):
  # Root items in the AST do not have to be declared before they can be used.
  # We therefore declare all names upfront before we resolve them.
  if isinstance(node, ast.Root):
    subscope = Scope(cannot_shadow=True, parent=scope)
    for child in node.children():
      declare_node(child, subscope)
    for child in node.children():
      resolve_node(child, subscope)
    return

  # Handle module and function declarations.
  if isinstance(node, ast.ModItem) or isinstance(node, ast.FnItem):
    subscope = Scope(parent=scope)
    for child in node.children():
      resolve_node(child, subscope)
      declare_node(child, subscope)
    return

  # Handle nodes with nested scopes.
  if isinstance(node, ast.IfStmt):
    resolve_node(node.cond, scope)
    subscope = Scope(parent=scope)
    for child in node.then_stmts:
      resolve_node(child, subscope)
      declare_node(child, subscope)
    subscope = Scope(parent=scope)
    for child in node.else_stmts:
      resolve_node(child, subscope)
      declare_node(child, subscope)
    return

  # Resolve children.
  for child in node.children():
    resolve_node(child, scope)

  # Handle specific nodes.
  if isinstance(node, ast.IdentExpr):
    node.binding.target = scope.resolve(node.name.spelling(), node.loc)

  if isinstance(node, ast.CallExpr):
    name = node.name.spelling()
    if name not in BUILTIN_FUNCS:
      node.binding.target = scope.resolve(name, node.name.loc)


# Declare the names of a given node.
def declare_node(node: ast.AstNode, scope: Scope):
  if isinstance(node, (
      ast.ModItem,
      ast.FnItem,
      ast.FnParam,
      ast.FnArg,
      ast.InputStmt,
      ast.OutputStmt,
      ast.LetStmt,
  )):
    scope.declare(node.name.spelling(), node)
    return

File: lib/Syntax/Names.cpp
//===----------------------------------------------------------------------===//
//
// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
// See https://llvm.org/LICENSE.txt for license information.
// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
//
//===----------------------------------------------------------------------===//

#include "silicon/LLVM.h"
#include "silicon/Syntax/AST.h"
#include "silicon/Syntax/Names.h"
#include "llvm/ADT/ScopeExit.h"

#define DEBUG_TYPE "names"

using namespace silicon;

/// Get the location of a binding.
static Location getLoc(ast::Binding binding) {
  return TypeSwitch<ast::Binding, Location>(binding)
      .Case<ast::FnItem *, ast::FnArg *, ast::LetStmt *>(
          [](auto *node) { return node->loc; });
}

/// Print a binding as `<type>(<address>)` for debugging purposes.
static llvm::raw_ostream &operator<<(llvm::raw_ostream &os,
                                     ast::Binding binding) {
  TypeSwitch<ast::Binding>(binding)
      .Case<ast::FnItem *, ast::FnArg *, ast::LetStmt *>([&](auto *node) {
        os << node->getTypeName() << "(" << static_cast<void *>(node) << ")";
      });
  return os;
}

namespace {
struct Declare {};
struct Resolve {};

struct Scope {
  SmallDenseMap<StringRef, ast::Binding> names;
};

/// A visitor that walks the AST and resolves names.
struct Resolver : public ast::Visitor<Resolver> {
  using Visitor::visit;
  using Visitor::visitNode;
  SmallVector<Scope> scopes;
  bool hasError = false;

  /// Declare a name in the current scope. If `redeclare` is true, it is
  /// allowed to redeclare an existing name, otherwise an error is emitted.
  void declareName(StringRef name, ast::Binding node, bool redeclare) {
    auto &slot = scopes.back().names[name];
    if (slot && !redeclare) {
      auto nodeLoc = getLoc(node);
      auto slotLoc = getLoc(slot);
      auto d = mlir::emitError(nodeLoc)
               << "name `" << name << "` already defined";
      d.attachNote(slotLoc)
          << "previous definition of `" << name << "` was here";
      hasError = true;
      return;
    }
    LLVM_DEBUG(dbgs().indent(scopes.size() * 2)
               << (slot ? "Red" : "D") << "eclaring `" << name << "` as "
               << node << " " << getLoc(node) << "\n");
    slot = node;
  }

  /// Resolve a name in the current scope or one of its parents. Reports an
  /// error for the given location if the name cannot be found.
  ast::Binding resolveName(StringRef name, Location loc) {
    for (auto &scope : llvm::reverse(scopes)) {
      if (auto node = scope.names.lookup(name)) {
        LLVM_DEBUG(dbgs().indent(scopes.size() * 2)
                   << "Resolved `" << name << "` to " << node << " " << loc
                   << "\n");
        return node;
      }
    }
    mlir::emitError(loc) << "unknown name `" << name << "`";
    hasError = true;
    return {};
  }

  /// Create a new scope and return a scope exit guard that will pop it when
  /// it goes out of scope (pun intended).
  auto makeScope() {
    LLVM_DEBUG(dbgs().indent(scopes.size() * 2) << "Subscope\n");
    scopes.emplace_back();
    return llvm::make_scope_exit([this] { scopes.pop_back(); });
  }

  // Do not descend into child nodes when declaring names. We only want to
  // collect order-agnostic items in this phase.
  template <typename T>
  void visitNode(T &node, Declare) {}

  // Declare names for order-agnostic items.
  void visitNode(ast::FnItem &item, Declare) {
    declareName(item.name, &item, false);
  }

  /// Standard handling for AST nodes that create a subscope for everything they
  /// contain.
  template <typename T>
  void resolveScope(T &node) {
    auto guard = makeScope();
    node.walk(*this, Declare{});
    if (hasError)
      return;
    node.walk(*this, Resolve{});
  }
  void visitNode(ast::Root &node, Resolve) { resolveScope(node); }
  void visitNode(ast::FnItem &node, Resolve) { resolveScope(node); }

  // Declare and resolve names.
  void visitNode(ast::FnArg &node, Resolve) {
    node.walk(*this, Resolve{});
    declareName(node.name, &node, false);
  }
  void visitNode(ast::LetStmt &node, Resolve) {
    node.walk(*this, Resolve{});
    declareName(node.name, &node, true);
  }
  void visitNode(ast::IdentExpr &node, Resolve) {
    node.binding = resolveName(node.name, node.loc);
  }
};
} // namespace

LogicalResult silicon::resolveNames(AST &ast) {
  Resolver resolver;
  ast.walk(resolver, Resolve{});
  return failure(resolver.hasError);
}
